### PR TITLE
executor, sessionctx, util: treat decimal truncation as warning during inserting

### PR DESCRIPTION
**Pull request by zz-jason**
fix https://github.com/pingcap/tidb/issues/3895

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @zimulala PTAL

File: util/types/datum.go [hunk 1]
```diff
@@ -1120,7 +1120,13 @@ func ProduceDecWithSpecifiedTp(dec *MyDecimal, tp *FieldType, sc *variable.State
 		} else if frac != decimal {
 			dec.Round(dec, decimal, ModeHalfEven)
 			if !dec.IsZero() && frac > decimal {
-				err = sc.HandleTruncate(ErrTruncated)
+				if sc.InInsertStmt {
```
Comment (jackysp) [inline util/types/datum.go]: Why don't check it inside `HandleTruncate`?

Comment (zz-jason) [inline util/types/datum.go]: in `HandleTruncate`, we don't know whether this is a decimal column.

Comment (jackysp) [inline util/types/datum.go]: So it is a decimal special issue, got it.

**Comment by XuHuaiyu** [comment]
LGTM

File: executor/executor_test.go [hunk 2]
```diff
@@ -2088,6 +2088,21 @@ func (s *testSuite) TestTiDBCurrentTS(c *C) {
 	c.Assert(terror.ErrorEqual(err, variable.ErrReadOnly), IsTrue)
 }
 
+func (s *testSuite) TestInsertDecimalTruncated(c *C) {
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: you can put this test in `executor/write_test.go:34:func (s *testSuite) TestInsert(c *C) `

Comment (zz-jason) [inline executor/executor_test.go]: moved

File: executor/write_test.go [hunk 3]
```diff
@@ -175,6 +175,15 @@ func (s *testSuite) TestInsert(c *C) {
 	tk.MustExec("create table t1 (b char(0));")
 	_, err = tk.Exec(`insert into t1 values ("");`)
 	c.Assert(err, IsNil)
+
+	// issue 3895
+	tk = testkit.NewTestKit(c, s.store)
+	tk.MustExec("USE test;")
+	tk.MustExec("DROP TABLE IF EXISTS t;")
+	tk.MustExec("CREATE TABLE t(a DECIMAL(4,2));")
+	tk.MustExec("INSERT INTO t SELECT CAST(1.000 AS DECIMAL(4,2));")
```
Comment (zimulala) [inline executor/write_test.go]: Plase add the test like `insert t values(1.000000000);`.

Comment (zz-jason) [inline executor/write_test.go]: added, PTAL

File: executor/write_test.go [hunk 4]
```diff
@@ -184,6 +184,9 @@ func (s *testSuite) TestInsert(c *C) {
 	tk.MustExec("INSERT INTO t SELECT CAST(1.000 AS DECIMAL(4,2));")
 	result := tk.MustQuery("SHOW WARNINGS;")
 	result.Check(testkit.Rows("Warning 1265 Data Truncated"))
+	tk.MustExec("INSERT INTO t VALUES (1.000000);")
+	result = tk.MustQuery("SHOW WARNINGS;")
+	result.Check(testkit.Rows("Warning 1265 Data Truncated"))
```
Comment (zimulala) [inline executor/write_test.go]: Can we add a `TODO` here?

**Comment by zz-jason** [comment]
@zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**